File: 2018 MCMC/monte_carlo_sim.py
# -*- coding: utf-8 -*-
"""
Created on Wed Jan  8 19:38:58 2020

@author: NTellaku
"""

# -*- coding: utf-8 -*-
"""
Created on Wed Jan  8 14:22:40 2020

@author: NTellaku
"""

import requests
import pandas as pd
import numpy as np
import scipy.stats as stats
import seaborn as sns
import matplotlib.pyplot as plt

def team_win_loss(dataset, i):
    """
    Adding the win-loss record
    dataset: which dataset to use?
    i: iterator
    """
    
    team = dataset[dataset["Team"] == team_ids[i]]
    team_wins = sum(n > 0 for n in team["Margin"])
    team_ties = sum(n == 0 for n in team["Margin"])
    team_loss = sum(n < 0 for n in team["Margin"])
    points = np.sum(team.PointsFor)
    wl_info = pd.DataFrame([[team_ids[i], team_wins, team_loss,
                             team_ties, points]],
                           columns = ["Team", "Wins", "Losses", "Ties", "Points"])
    return wl_info

def matchup(home_team, away_team, week):
    """
    Input:  home_team: home team number
            away_team: away team number
            week: which week for the matchup
            
    calculates who wins each individual matchup.
    assuming each team scores assumes a gaussian distribution
    """
    
    #getting the summaries for each team
    summary_stats = (pd.DataFrame([analysis.sort_values(by = ["Team"])
                                           .Team
                                           .unique(),
                                   analysis.groupby("Team")
                                           .mean()["PointsFor"],
                                   analysis.groupby("Team")
                                           .std()["PointsFor"]]).T)
    
    #cleaning the data set
    colnames = ["Team", "AvgPoints", "StDPoints"]
    summary_stats.columns = colnames
    
    #find each team's scores
    home = np.random.normal(summary_stats[summary_stats.Team == home_team].AvgPoints,
                            summary_stats[summary_stats.Team == away_team].StDPoints,
                            1) #only one score "Any given Sunday"
    away = np.random.normal(summary_stats[summary_stats.Team == home_team].AvgPoints,
                            summary_stats[summary_stats.Team == away_team].StDPoints,
                            1)
    
    #make same format as analysis data set
    home_row = np.array([week, home_team, home[0], (home - away)[0], away[0]])
    away_row = np.array([week, away_team, away[0], (away - home)[0], home[0]])
    
    #apped these rows to the analysis data set
    extra_matchup = (pd.DataFrame([home_row, away_row],
                                  columns = analysis.columns))
    
    return extra_matchup

#team number to name Dictionary
mapping = {1: "Mount",
           2: "Alec",
           3: "Sirpi",
           4: "Oatman",
           5: "Babcock",
           9: "Jordan",
           11: "Casey",
           12: "Badillo",
           13: "Naki",
           14: "Kooper"}


league_id = 298982
year = 2018


url = "https://fantasy.espn.com/apis/v3/games/ffl/leagueHistory/" + str(league_id) + "?seasonId=" + str(year)
r = requests.get(url, params = {"view": "mMatchup"})
d = r.json()[0]

maximum = max(pd.DataFrame(d["schedule"]).index.tolist()) #how many obs?
length_df = pd.DataFrame([[d["schedule"][i]["winner"]] for i in range(maximum)])
length_df = length_df[length_df[0] != "UNDECIDED"]
length = range(len(length_df))


#Selecting weeks, points, data
source = pd.DataFrame([[d["schedule"][i]["matchupPeriodId"],
                        d["schedule"][i]["home"]["teamId"],
                        d["schedule"][i]["home"]["totalPoints"],
                        d["schedule"][i]["away"]["teamId"],
                        d["schedule"][i]["away"]["totalPoints"]] for i in length],
                        columns = ["Week", "Team1", "Score1", "Team2", "Score2"])

#add margin of defeat/victory
margins = source.assign(Margin1 = source["Score1"] - source["Score2"],
                        Margin2 = source["Score2"] - source["Score1"])

#only first 9 weeks of data
source_9 = margins[margins.Week < 10]

#transform from wide to long
source_long = (source_9.loc[:, ("Week", "Team1", "Score1", "Margin1")]
                         .rename(columns = {"Team1": "Team",
                                            "Score1": "Score",
                                            "Margin1": "Margin"})
                         .append(source_9.loc[:, ("Week", "Team2", "Score2", "Margin2")]
                             .rename(columns = {"Team2": "Team",
                                                "Score2": "Score",
                                                "Margin2": "Margin"})))

source_long = (source_long.assign(PA = source_long.Score - source_long.Margin)
                          .rename(columns = {"Score": "PointsFor",
                                             "PA": "Points Against"}))


#simming 1000 times
sim_standings = pd.DataFrame()

for l in range(1000):

    analysis = source_long.copy()
    
    #Outer loop to create the weekly object
    for i in range(10, 14):
        weekly_matchups = margins[margins.Week == i].loc[:, ["Team1", "Team2"]]
        week = i
        
        #inner loop to do the matchup
        for j in range(5):
            home_team = weekly_matchups.iloc[j, 0]
            away_team = weekly_matchups.iloc[j, 1]
            result = matchup(home_team, away_team, i)
            analysis = analysis.append(result)
            
            
    #creating record from values
    team_ids = source_long.Team.unique()
    
    #initialize an empty dataframe to append to
    win_loss = []
    
    #loop through all the teams and have the rows append
    for j in range(len(team_ids)):
        row = team_win_loss(analysis, j)
        win_loss.append(row)
    
    win_loss = (pd.concat(win_loss)
                  .sort_values(by = ["Wins", "Ties", "Points"], ascending = False)
                  .assign(Standing = np.arange(1, 11))
                  .sort_values(by = ["Team"])
                  .reset_index(drop = True))
    
    sim_standings[l] = win_loss.Standing

#keeping a separate copy to avoid computation time
sim_standings2 = sim_standings.copy()    

sim_standings.insert(loc = 0, value = win_loss.Team, column = "Team")

#adding team name, making it the index
final = sim_standings.replace({"Team": mapping}).set_index("Team")

#how many at each rank?
final.loc["Mount", :].value_counts()
counts = pd.DataFrame([final.loc[i, :].value_counts() for i in mapping.values()])

#replace NA with 0
counts.fillna(0, inplace = True)

#where are the teams most likely to place based on the sim?
predicted_ranks = (pd.DataFrame([counts.max(), counts.idxmax()])
                        .T
                        .assign(Rank = np.arange(1, 11)))
colnames = ["Count", "Team", "Rank"]
predicted_ranks.columns = colnames
predicted_ranks = predicted_ranks.set_index("Team")


#Observed ranks
#transform from wide to long
source_13 = margins[margins.Week <= 13]
margins_long = (source_13.loc[:, ("Week", "Team1", "Score1", "Margin1")]
                         .rename(columns = {"Team1": "Team",
                                            "Score1": "PointsFor",
                                            "Margin1": "Margin"})
                         .append(source_13.loc[:, ("Week", "Team2", "Score2", "Margin2")]
                             .rename(columns = {"Team2": "Team",
                                                "Score2": "PointsFor",
                                                "Margin2": "Margin"})))

#initialize an empty dataframe to append to
final_win_loss = []

#loop through all the teams and have the rows append
for j in range(len(team_ids)):
    row = team_win_loss(margins_long, j)
    final_win_loss.append(row)

final_win_loss = (pd.concat(final_win_loss)
                    .sort_values(by = ["Wins", "Ties", "Points"], ascending = False)
                    .assign(Standing = np.arange(1, 11))
                    .sort_values(by = ["Team"])
                    .reset_index(drop = True))

comparison = (final_win_loss.replace({"Team": mapping})
                            .set_index("Team")
                            .merge(how = "right",
                                   right = predicted_ranks,
                                   left_index = True,
                                   right_index = True)
                            .loc[:, ["Rank", "Standing"]]
                            .rename(columns = {"Rank": "Predicted",
                                               "Standing": "Observed"}))
tau, p_value = stats.kendalltau(comparison.Predicted,
                                comparison.Observed)


#heatmap of the percentages
counts2 = counts.reindex(predicted_ranks.index) / 10
cmap = sns.diverging_palette(10, 150, as_cmap = True)
sns.heatmap(counts2, cmap = cmap, cbar = False,
            annot = counts2, linewidth = 0.5)
plt.title("1000 Simulations - Percentage of Each Final Standing")

File: 2018 MCMC/monte_carlo_sim.R
library(tidyverse)

nikhil <- c(130, 101, 83, 127, 90, 125, 93, 97, 105)
mount <- c(76, 69, 97, 70, 85, 111, 98, 63, 77)
alec <- c(71, 84, 59, 103, 99, 85, 140, 84, 99)
sirpi <- c(102, 87, 69, 53, 79, 96, 75, 99, 64)
oatman <- c(107, 78, 89, 115, 96, 139, 91, 118, 91)
babcock <- c(87, 100, 106, 85, 93, 91, 118, 89, 124)
jordan <- c(139, 98, 118, 137, 80, 67, 98, 79, 99)
casey <- c(97, 103, 93, 142, 113, 123, 120, 149, 134)
badillo <- c(98, 128, 81, 97, 71, 76, 77, 111, 104)
kooper <- c(89, 90, 74, 98, 102, 124, 77, 95, 102)

name <- c("Nikhil", "Mount", "Alec", "Sirpi", "Oatman", "Babcock", 
          "Jordan", "Casey", "Badillo", "Kooper")

temp <- as.tibble(rbind(nikhil, mount, alec, sirpi, oatman,
                        babcock, jordan, casey, badillo, kooper))

temp <- temp %>% mutate(mean = apply(temp, 1, mean), sd = apply(temp, 1, sd), 
                        pfor = apply(temp, 1, sum), name = name)

wins <- c(7, 3, 5, 2, 6, 3, 5, 6, 5, 3)
loss <- length(nikhil) - wins
pagt <- c(865, 876, 790, 850, 830, 926, 868, 964, 839, 937)

league_info <- as.tibble(cbind(temp$name, wins, loss, temp$pfor, pagt,
                               temp$mean, temp$sd)) %>% 
  rename("Team" = V1, "Wins" = wins, "Losses" = loss, "Points For" = V2,
         "Points Against" = pagt, "Mean" = V3, "Standard Deviation" = V4) %>% 
  mutate(new_wins = wins, new_losses = loss)


#Function to see who wins the matchup
matchup <- function(sims, team1, team2) {
  
  #sims: number of simulations
  #team1: The first team in the matchup
  #team2: The second team in the matchup
  
  home <- rnorm(sims, as.numeric(league_info %>%
                                filter(Team == team1) %>%
                                select(Mean)),
                as.numeric(league_info %>%
                             filter(Team == team1) %>%
                             select("Standard Deviation")))
  away <- rnorm(sims, as.numeric(league_info %>%
                                filter(Team == team2) %>%
                                select(Mean)),
                as.numeric(league_info %>%
                             filter(Team == team2) %>%
                             select("Standard Deviation")))
  
  #for loop to determine who has more higher scores
  
  result = 0 #initialize
  for (i in c(1:sims)) {
    if (home[i] > away[i]) {
      result[i] = 1
    } else {
      result[i] = 0
    }
  }
  
  #checking if simulations is odd or even for conditional, want odd
  if ((sims %% 2) == 0) {
    sims = sims + 1
  }
  
  #conditional to add the win and loss statistic
  if (sum(result) >= ceiling(sims / 2)) {
    a <- league_info %>% 
      filter(Team == team1) %>% 
      mutate(new_wins = new_wins + 1)
    
    b <- league_info %>%
      filter(Team == team2) %>%
      mutate(new_losses = new_losses + 1)
  } else {
    a <- league_info %>%
      filter(Team == team1) %>%
      mutate(new_losses = new_losses + 1)
    
    b <- league_info %>%
      filter(Team == team2) %>%
      mutate(new_wins = new_wins + 1)
  }
  c <- rbind(a, b)
  return(c)
}


#Running the program multiple times
output <- vector("list", 1000)

for (i in 1:1000) {
  
  league_info <- as.tibble(cbind(temp$name, wins, loss, temp$pfor, pagt,
                                 temp$mean, temp$sd)) %>% 
    rename("Team" = V1, "Wins" = wins, "Losses" = loss, "Points For" = V2,
           "Points Against" = pagt, "Mean" = V3, "Standard Deviation" = V4) %>% 
    mutate(new_wins = wins, new_losses = loss)
  
  #Week 9
  #league_info <- rbind(league_info,
  #                     matchup(1, "Nikhil", "Jordan"),
  #                     matchup(1, "Sirpi", "Oatman"),
  #                     matchup(1, "Casey", "Babcock"),
  #                     matchup(1, "Badillo", "Kooper"),
  #                     matchup(1, "Mount", "Alec")) %>% 
  #  slice((n() - 9):n())
  
  #Week 10
  league_info <- rbind(league_info,
                       matchup(1, "Nikhil", "Oatman"),
                       matchup(1, "Casey", "Sirpi"),
                       matchup(1, "Babcock", "Badillo"),
                       matchup(1, "Jordan", "Mount"),
                       matchup(1, "Kooper", "Alec")) %>% 
    slice((n() - 9):n())
  
  #Week 11
  league_info <- rbind(league_info,
                       matchup(1, "Nikhil", "Sirpi"),
                       matchup(1, "Badillo", "Casey"),
                       matchup(1, "Mount", "Oatman"),
                       matchup(1, "Alec", "Babcock"),
                       matchup(1, "Kooper", "Jordan")) %>% 
    slice((n() - 9):n())
  
  #Week 12
  league_info <- rbind(league_info,
                       matchup(1, "Nikhil", "Mount"),
                       matchup(1, "Badillo", "Sirpi"),
                       matchup(1, "Alec", "Casey"),
                       matchup(1, "Oatman", "Kooper"),
                       matchup(1, "Babcock", "Jordan")) %>% 
    slice((n() - 9):n())
  
  pfor <- as.numeric(league_info$`Points For`)
  
  league_info <- league_info %>% 
    mutate(pfor) %>% 
    select(-"Points For") %>% 
    rename("Points For" = pfor) %>% 
    arrange(desc(new_wins), desc(`Points For`)) %>% 
    mutate(rank = c(1:10))
  
  league_extra <- league_info %>% 
    select("Team", rank) %>% 
    arrange(Team, rank)
  
  output[i] <- league_extra %>% select(rank)
}

#slicing the list to just the number each player gets a rank
monte <- as.data.frame(output)
colnames(monte) <- rep("Rank", 1000)
monte$newcolumn <- league_extra$Team
monte <- monte[c(1001, 1:1000)]
monte <- t(monte)
colnames(monte) <- league_extra$Team
monte <- as.tibble(monte) %>% slice(2:n())

alec_rank <- t(as.data.frame(table(monte$Alec)))
colnames(alec_rank) <- alec_rank[1, ]
alec_rank <- as.data.frame(alec_rank)
alec_rank <- alec_rank[2, ]
row.names(alec_rank) <- "Alec"

babcock_rank <- t(as.data.frame(table(monte$Babcock)))
colnames(babcock_rank) <- babcock_rank[1, ]
babcock_rank <- as.data.frame(babcock_rank)
babcock_rank <- babcock_rank[2, ]
row.names(babcock_rank) <- "Babcock"

badillo_rank <- t(as.data.frame(table(monte$Badillo)))
colnames(badillo_rank) <- badillo_rank[1, ]
badillo_rank <- as.data.frame(badillo_rank)
badillo_rank <- badillo_rank[2, ]
row.names(badillo_rank) <- "Badillo"

casey_rank <- t(as.data.frame(table(monte$Casey)))
colnames(casey_rank) <- casey_rank[1, ]
casey_rank <- as.data.frame(casey_rank)
casey_rank <- casey_rank[2, ]
row.names(casey_rank) <- "Casey"

jordan_rank <- t(as.data.frame(table(monte$Jordan)))
colnames(jordan_rank) <- jordan_rank[1, ]
jordan_rank <- as.data.frame(jordan_rank)
jordan_rank <- jordan_rank[2, ]
row.names(jordan_rank) <- "Jordan"

kooper_rank <- t(as.data.frame(table(monte$Kooper)))
colnames(kooper_rank) <- kooper_rank[1, ]
kooper_rank <- as.data.frame(kooper_rank)
kooper_rank <- kooper_rank[2, ]
row.names(kooper_rank) <- "Kooper"

mount_rank <- t(as.data.frame(table(monte$Mount)))
colnames(mount_rank) <- mount_rank[1, ]
mount_rank <- as.data.frame(mount_rank)
mount_rank <- mount_rank[2, ]
row.names(mount_rank) <- "Mount"

nikhil_rank <- t(as.data.frame(table(monte$Nikhil)))
colnames(nikhil_rank) <- nikhil_rank[1, ]
nikhil_rank <- as.data.frame(nikhil_rank)
nikhil_rank <- nikhil_rank[2, ]
row.names(nikhil_rank) <- "Nikhil"

oatman_rank <- t(as.data.frame(table(monte$Oatman)))
colnames(oatman_rank) <- oatman_rank[1, ]
oatman_rank <- as.data.frame(oatman_rank)
oatman_rank <- oatman_rank[2, ]
row.names(oatman_rank) <- "Oatman"

sirpi_rank <- t(as.data.frame(table(monte$Sirpi)))
colnames(sirpi_rank) <- sirpi_rank[1, ]
sirpi_rank <- as.data.frame(sirpi_rank)
sirpi_rank <- sirpi_rank[2, ]
row.names(sirpi_rank) <- "Sirpi"

final <- as.tibble(Reduce(function(x, y) merge(x, y, all = TRUE),
                          list(alec_rank, babcock_rank, badillo_rank,
                               casey_rank, jordan_rank, kooper_rank,
                               mount_rank, nikhil_rank, oatman_rank,
                               sirpi_rank))) %>%
  select(8, 6, 7, 5, 10, 9, 1, 2, 3, 4)


the_names <- c("Alec", "Babcock", "Badillo", "Jordan", "Kooper", "Mount",
               "Sirpi", "Casey", "Nikhil", "Oatman")
rownames(final) <- the_names
